File: resources/objects/obj_bee.cpp
/*
* Copyright (c) 2015-18 Luke Montalvo <lukemontalvo@gmail.com>
*
* This file is part of BEE.
* BEE is free software and comes with ABSOLUTELY NO WARANTY.
* See LICENSE for more details.
*/

#ifndef RES_OBJ_BEE
#define RES_OBJ_BEE 1

#include "../resources.hpp"

#include "obj_bee.hpp"

ObjBee::ObjBee() : Object("obj_bee", "obj_bee.hpp") {
	implemented_events.insert({
		bee::E_EVENT::CREATE,
		bee::E_EVENT::DESTROY,
		bee::E_EVENT::COMMANDLINE_INPUT,
		bee::E_EVENT::PATH_END,
		bee::E_EVENT::OUTSIDE_ROOM,
		bee::E_EVENT::DRAW
	});
}
void ObjBee::create(bee::Instance* self) {
	self->set_computation_type(bee::E_COMPUTATION::SEMIPLAYER);
	self->get_physbody()->set_shape(bee::E_PHYS_SHAPE::BOX, new double[3] {100.0, 100.0, 100.0});
	self->get_physbody()->set_mass(0.0);
	self->get_physbody()->add_constraint(bee::E_PHYS_CONSTRAINT::FLAT, nullptr);

	// create event
	std::cout << "u r a b " << self->id << "\n";
	_p("text_id") = nullptr;

	bee::Instance* ui_handle = bee::ui::create_handle(self->get_corner().first, self->get_corner().second, 100, 100, self);
	ObjUIHandle* obj_ui_handle = static_cast<ObjUIHandle*>(ui_handle->get_object());
	obj_ui_handle->set_is_visible(ui_handle, false);
	_p("ui_handle") = ui_handle;

	if (self == obj_bee->get_instance(0)) {
		_a("serialdata") = self->serialize();
		_p("path") = nullptr;
		_d("path_previous_mass") = 0.0;

		bee::kb::bind(SDLK_UNKNOWN, bee::KeyBind("StartPath", [this, self] (const SDL_Event* e) mutable {
			self = obj_bee->get_instance(0);
			s = &self->get_data();

			if (_p("path") != nullptr) {
				delete static_cast<bee::PathFollower*>(_p("path"));
				_p("path") = nullptr;
			}

			self->set_mass(1.0);
			bee::PathFollower* pf = new bee::PathFollower(path_bee, {100.0, 100.0, 0.0}, 100);
			pf->is_curved = true;
			bee::get_current_room()->automate_path(self, pf);
			_p("path") = pf;
		}));

		bee::kb::bind(SDLK_UNKNOWN, bee::KeyBind("StartSerialize", [this, self] (const SDL_Event* e) mutable {
			self = obj_bee->get_instance(0);
			s = &self->get_data();

			_m("serialdata") = self->serialize();
			bee::messenger::send({"bee"}, bee::E_MESSAGE::INFO, _a("serialdata").to_str(true));
		}));
		bee::kb::bind(SDLK_UNKNOWN, bee::KeyBind("StartDeserialize", [this, self] (const SDL_Event* e) mutable {
			self = obj_bee->get_instance(0);
			s = &self->get_data();

			self->deserialize(_m("serialdata"));
		}));
	}
}
void ObjBee::destroy(bee::Instance* self) {
	if (self == obj_bee->get_instance(0)) {
		_a("serialdata").reset();

		if (_p("path") != nullptr) {
			delete static_cast<bee::PathFollower*>(_p("path"));
			_p("path") = nullptr;
		}
	}

	delete static_cast<bee::TextData*>(_p("text_id"));

	bee::ui::destroy_parent(self);

	Object::destroy(self);
}
void ObjBee::commandline_input(bee::Instance* self, const std::string& str) {
	std::cout << "bee" << self->id << ":~~~" << str << "~~~\n";
}
void ObjBee::path_end(bee::Instance* self, bee::PathFollower* pf) {
	bee::get_current_room()->automate_path(self, nullptr);

	delete static_cast<bee::PathFollower*>(_p("path"));
	_p("path") = nullptr;
}
void ObjBee::outside_room(bee::Instance* self) {
	bee::get_current_room()->destroy(self);
}
void ObjBee::draw(bee::Instance* self) {
	int size = 100;
	double r = util::radtodeg(self->get_physbody()->get_rotation_z());
	self->draw(size, size, r, bee::RGBA(bee::E_RGB::WHITE));

	if (static_cast<bee::PathFollower*>(_p("path")) != nullptr) {
		path_bee->draw(static_cast<bee::PathFollower*>(_p("path")));
	}

	_p("text_id") = font_liberation->draw(static_cast<bee::TextData*>(_p("text_id")), self->get_corner().first, self->get_corner().second, std::to_string(self->id));

	lt_bee->set_position(glm::vec4(self->get_x(), self->get_y(), 0.0, 1.0));
	lt_bee->set_color({static_cast<Uint8>(self->id*50), static_cast<Uint8>(self->id*20), 255, 255});
	lt_bee->queue();
}

#endif // RES_OBJ_BEE

File: maps/lv_test/resources/objects/obj_bee.py
###############################################################################
# Copyright (c) 2015-18 Luke Montalvo <lukemontalvo@gmail.com>
#
# This file is part of BEE.
# BEE is free software and comes with ABSOLUTELY NO WARRANTY
# See LICENSE for more details.
###############################################################################

import bee

def create(self):
    self.set_computation_type(bee.E_COMPUTATION["SEMIPLAYER"])
    pb = self.get_physbody()
    pb.set_shape(bee.E_PHYS_SHAPE["BOX"], (100.0, 100.0, 100.0))
    pb.set_mass(0.0)
    pb.add_constraint(bee.E_PHYS_CONSTRAINT["FLAT"], ())

    print("u r a b {}".format(self.get_id()))

    ui_handle = bee.ui.create_handle(int(self.get_corner()[0]), int(self.get_corner()[1]), 100, 100, self)
    ui_handle.set_data("is_visible", False)
    self.set_data("ui_handle", ui_handle)

    if self == bee.Instance("obj_bee").at(0):
        bee.console.set_var("inst_bee_serialdata", self.serialize())

        bee.kb.bind("SDLK_UNKNOWN", {
            "name": "StartPath",
            "func": lambda e: (
                self.set_mass(1.0),
                bee.get_current_room().automate_path(self, {
                    "path": bee.Path("path_bee"),
                    "offset": (100.0, 100.0, 0.0),
                    "speed": 100,
                    "is_curved": True
                })
            )
        })
        bee.kb.bind("SDLK_UNKNOWN", {
            "name": "StartTimeline",
            "func": lambda e: (
                bee.get_current_room().automate_timeline({
                    "timeline": bee.Timeline("tl_bee"),
                    "start_offset": 0,
                    "is_looping": False,
                    "is_pausable": False
                })
            )
        })

        bee.kb.bind("SDLK_UNKNOWN", {
            "name": "StartSerialize",
            "func": lambda e: (
                bee.console.set_var("inst_bee_serialdata", self.serialize()),
                bee.messenger.send(["obj_bee"], bee.E_MESSAGE["INFO"], str(bee.console.get_var("inst_bee_serialdata")))
            )
        })
        bee.kb.bind("SDLK_UNKNOWN", {
            "name": "StartDeserialize",
            "func": lambda e: (
                self.deserialize(bee.console.get_var("inst_bee_serialdata"))
            )
        })

def destroy(self):
    if self == bee.Instance("obj_bee").at(0):
        bee.console.set_var("inst_bee_serialdata", None)

    bee.ui.destroy_parent(self)

def commandline_input(self, input):
    print("bee{}:~~~{}~~~".format(self.get_id(), input))
def path_end(self, pf):
    bee.messenger.log("path end")
    bee.get_current_room().automate_path(self, None)
def outside_room(self):
    bee.get_current_room().destroy(self)

def draw(self):
    import math
    size = 100
    r = math.degrees(self.get_physbody().get_rotation()[2])
    self.draw(size, size, r, bee.E_RGBA["WHITE"])

    paths = bee.get_current_room().get_paths()
    if self in paths:
        pf = paths[self]
        if pf["path"]:
            pf["path"].draw(pf)

    bee.Font("font_liberation").draw_fast(int(self.get_corner()[0]), int(self.get_corner()[1]), str(self.get_id()))

    l = bee.Light("lt_bee")
    c = l.get_color()
    l.set_position((self.get_pos()[0], self.get_pos()[1], 0.0, 1.0))
    l.set_color((self.get_id()*50, self.get_id()*20, c[2], c[3]))
    l.queue()
